src/tests/JIT/Directed/Convert/out_of_range_fp_to_int_conversions.cpp

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <math.h>
+
+#ifdef _MSC_VER
+#define DLLEXPORT __declspec(dllexport)
+#else
+#define DLLEXPORT __attribute__((visibility("default")))
+#endif
+
+typedef enum {
+    CONVERT_BACKWARD_COMPATIBLE,
+    CONVERT_SENTINEL,
+    CONVERT_SATURATING,
+    CONVERT_NATIVECOMPILERBEHAVIOR,
+    CONVERT_MANAGED_BACKWARD_COMPATIBLE_X86_X64,
+    CONVERT_MANAGED_BACKWARD_COMPATIBLE_ARM32,
+} FPtoIntegerConversionType;
+
+extern "C" DLLEXPORT int32_t ConvertDoubleToLong(double x, FPtoIntegerConversionType t)
+{
+    if (t == CONVERT_NATIVECOMPILERBEHAVIOR)
+        return (int32_t)x;
+
+    x = trunc(x); // truncate (round toward zero)
+
+    switch (t) {
+    case CONVERT_BACKWARD_COMPATIBLE:
+    case CONVERT_MANAGED_BACKWARD_COMPATIBLE_X86_X64:
+    case CONVERT_SENTINEL:
+        return ((x != x) || (x < INT32_MIN) || (x > INT32_MAX)) ? INT32_MIN : (int32_t)x;
+
+    case CONVERT_MANAGED_BACKWARD_COMPATIBLE_ARM32:
+    case CONVERT_SATURATING:
+        return (x != x) ? 0 : (x < INT32_MIN) ? INT32_MIN : (x > INT32_MAX) ? INT32_MAX : (int32_t)x;
+    case CONVERT_NATIVECOMPILERBEHAVIOR: // handled above, but add case to silence warning
+        return 0;
+    }
+
+    return 0;
+}
+
+extern "C" DLLEXPORT uint32_t ConvertDoubleToUnsignedLong(double x, FPtoIntegerConversionType t)
+{
+    if (t == CONVERT_NATIVECOMPILERBEHAVIOR)
+        return (uint32_t)x;
+
+    x = trunc(x); // truncate (round toward zero)
+    const double int64_max_plus_1 = 0x1.p63; // 0x43e0000000000000 // (uint64_t)INT64_MAX + 1;
+
+    switch (t) {
+    case CONVERT_MANAGED_BACKWARD_COMPATIBLE_X86_X64:
+    case CONVERT_BACKWARD_COMPATIBLE:
+        return ((x != x) || (x < INT64_MIN) || (x >= int64_max_plus_1)) ? 0 : (uint32_t)(int64_t)x;
```

Comment:
Is this problematic on platforms where the conversion to `int64_t` will return the sentinel vs saturate?

That is, say `x` is `-1`, this will become `-2` (`0xFFFFFFFFFFFFFFFE`) and then truncate to `4294967294` (`0xFFFFFFFE`)...

Reply:
I don't believe so. The conversion to int64_t here, is well defined on all platforms, and the int64_t to uint32_t cast is also well defined as just truncating on all platforms.